### ISSUE TITLE
require('vanilla-lazyload') returns "not a constructor"

### ISSUE THREAD

**Issue by lewishowles** [verlok/lazyload#236]
I have two projects that are copies of each other. For some reason, I'm getting two different values when doing `require('vanilla-lazyload')` - even though both use the same version (10.15.0).

I'm obviously missing something.

The docs mention different versions of the dist, but I don't know how to target a particular version. The module is installed via npm (Yarn), and I'm using Webpack to build.

Website 1:
```javascript
window.LazyLoad = require('vanilla-lazyload')
window.LazyLoadInstance = new LazyLoad({
	class_loading : 'lazy-loading',
	class_loaded  : 'lazy-loaded',
	class_error   : 'lazy-error',
	class_initial : 'lazy-initial',
});

console.log(LazyLoad) // function L()...
```

Website 2:
```javascript
window.LazyLoad = require('vanilla-lazyload')
window.LazyLoadInstance = new LazyLoad({
	class_loading : 'lazy-loading',
	class_loaded  : 'lazy-loaded',
	class_error   : 'lazy-error',
	class_initial : 'lazy-initial',
});

// Error, LazyLoad is not a constructor
console.log(LazyLoad) // Object { default: LazyLoad(), … }
```

Using `require('vanilla-lazyload').default` works on the second project, but I can't see why it would be different.

I have tried reinstalling the module on both, but that made no difference.

Any ideas?

**Comment by verlok** [comment]
Hey, Lewis,
it seems to me that your secont project has installed a version before 10.15, which fixed the “not a constructor” problem. 
Can you do a check for me? Look inside the `node_modules/vanilla-lazyload`

**Comment by verlok** [comment]
...and open the `package.json` file inside that folder. What is there inside the `version` field?

**Comment by verlok** [comment]
In case you have two different versions installed, run `npm install --save vanilla-lazyload@10.15.0`. This should solve it all.

**Comment by lewishowles** [comment]
Hi, thanks for the suggestion.

Website 1 is actually showing 10.11, while website 2 (the one with the constructor error) is 10.15 in its package.json, so that's odd.

I wiped my node_modules on website 1, reinstalled and LazyLoad is now showing 10.15.0 in its package.json, but it's now saying LazyLoad is not a constructor.

Is `require('vanilla-lazyload').default` how I should address it now?

**Comment by verlok** [comment]
Hi @lewishowles,
since version 10.15 LazyLoad exports itself as a ES6 module, so you can now import it both:

- from the HTML as `<script type="module"`> --> see repo [lazyload-es2015-test](https://github.com/verlok/lazyload-es2015-test) for reference
- using the ES6 module syntax `import LazyLoad from "vanilla-lazyload"` then bundle with WebPack --> see repo [lazyload-es2015-webpack-test](https://github.com/verlok/lazyload-es2015-webpack-test) for reference

I don't know exactly what `require('module').default` does, but from what I read on StackOverflow this might be the solution if you import the script using `require`.

Does it work for you?
And if yes... do you know why? 😄

**Comment by lewishowles** [comment]
It does work by using `require('module').default` in this case - I don't know why though, I just inspected what was returned when using require and noticed that the `.default` function of the object was the same as the main LazyLoad function!

I use Laravel Mix which uses webpack by default (I think), so your ES6 module syntax example also works.

Thanks for the great plugin by the way!

**Comment by verlok** [comment]
Maybe it's just the way Laravel Mix configures webpack in one project and in the other.
ES6 module syntax is best, I think, but the script should work with CommonJS too. 

I'll keep investigating.

You're welcome! 
If you feel like buying me a coffee, I'm [open to it](https://www.paypal.me/verlok) 😄😇

**Comment by verlok** [comment]
Hey @lewishowles,
I was able to reproduce the problem. 

<img width="1057" alt="LazyLoad is not a constructor" src="https://user-images.githubusercontent.com/1127721/44134467-631ff150-a065-11e8-9a9c-d9f69868e544.png">

It's caused by the way LazyLoad is built and the files exposed in its `package.json`.
I'm working on a fix right now.

In the meantime, if you can, just import LazyLoad using the es6 syntax.

```js
import LazyLoad from "path/to/lazyload.es2015.js"
```

I'll be back soon with a fix for the CommonJS (`require`) version.
Thanks for not

**Comment by verlok** [comment]
I can't find a way to tell WebPack to use the UMD version of LazyLoad instead of the ES6 modules version.

What's in LazyLoad's `package.json` file is:

```json
{
	"name": "vanilla-lazyload",
	"version": "10.15.0",
	"main": "dist/lazyload.min.js",
	"module": "dist/lazyload.es2015.js"
}
```

I also tried with `browser` instead of `module` as stated in the [package.json documentation](https://docs.npmjs.com/files/package.json).

`main` points to the UMD version (also tried with a pure commonJS version), whereas `main` points to the ES6 module version. 

But I got the same result: when bundling with webpack, it uses the ES6 version instead of the UMD/CommonJS one.

**I'm officially asking the GitHub community help here!** 😄 

How can I make WebPack use the file in the `main` field instead of the one in the `module`/`browser` field?

/cc @G100g

**Comment by verlok** [comment]
Hi @lewishowles,
I found out what to do to solve your issue with WebPack and LazyLoad.

The best solution which would work without touching the WebPack configuration would be to use the ES6 module syntax (aka ESM), and import LazyLoad like the following:

```js
import LazyLoad from "vanilla-lazyload"
```

If using the ESM syntax isn't possible for you, or for some reason you wish to continue using the CJS `require` syntax like you already do (`window.LazyLoad = require('vanilla-lazyload');`), you should configure WebPack to load the scripts from the package.json's `main` fields instead. See [configure webpack to resolve main fields](https://webpack.js.org/configuration/resolve/#resolve-mainfields).

Long story short, you should edit your `webpack.config.js` file and add/edit the following setting:

```json
module.exports = {
	/* Other settings */
	resolve: {
		mainFields: ["main", "browser", "module"]
	}
};
```

This tells WebPack to resolve your `require`s to the packages `main` field, which must contain contains the UMD version of the installed modules. [More info here](http://2ality.com/2017/04/setting-up-multi-platform-packages.html).

Please let me know if this solves for you.
Thanks

**Comment by verlok** [comment]
Closing this after 1 week. 
If you need any more information, feel free to ask.
Thanks
